### PR TITLE
[[ WidgetImpExp ]] Added ability to import / export widget state as a…

### PR DESCRIPTION
**Pull request by runrevmark**
You can now do:
  export <widget> to array tVar
and
  import widget from array tVar

The former will create an array with two keys $kind and $state representing the widget's internal state.

The latter will create a widget from such a widget internal state array.

### CODE REVIEW

**Comment by runrevmark** [comment]
I just realised - I need to add the mapping from LCB type-world to LCS type-world to this!

**Comment by livecode-vulcan** [comment]
📝 Hi @runrevmark, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.

**Comment by montegoulding** [comment]
@runrevmark looking forward to this being merged in. If I read bind right if the kind isn't loaded the state will be retained anyway so that at some point when the kind is available it can be initialised with the correct state. Is that correct?

**Comment by livecodefraser** [comment]
What still needs to be done with this PR?

**Comment by runrevmark** [comment]
@montegoulding: That's correct. If the engine doesn't know the type of widget at load time, then it won't instantiate an instance of the widget but will retain the state array which it was saved with. Although it doesn't happen yet, at some point the engine will go through and automagically instantiate any widgets when the appropriate extension is loaded. At the moment, if an appropriate extension isn't loaded at the point the widget is, it will remain inert until the stack is reloaded.

(The import from array command does exactly the same as when loading the object from a stackfile - so you can use it to create widgets which are 'unknown type', the given state will be preserved).

**Comment by runrevmark** [comment]
@livecodefraser: I think it is okay as it is for now. It isn't quite finished on two fronts - deciding whether $state and $kind are the correct way to go to annotate the array and extending support to all object types. The main reason for adding this at the moment was so that @montegoulding could get lcVCS working with widgets, a purpose for which this should be sufficient.

File: engine/src/cmdsf.cpp [hunk 1]
```diff
@@ -2734,7 +2820,11 @@ void MCImport::compile(MCSyntaxFactoryRef ctxt)
 			MCSyntaxFactoryExecMethod(ctxt, kMCInterfaceExecImportSnapshotOfScreenMethodInfo);
 		}
 	}
-	else
+	else if (format == EX_OBJECT)
+    {
+        // COMPILE-TODO
```
Comment (livecodefraser) [inline engine/src/cmdsf.cpp:2825]: Is there something we can do here to indicate an error?

Comment (runrevmark) [inline engine/src/cmdsf.cpp:2825]: That code isn't used at the moment - it is part of the 'compile' stuff for later use when translating scripts.

**Comment by livecodefraser** [comment]
@livecode-vulcan review okay d0fcd78

**Comment by livecode-vulcan** [comment]
📝 Hi @runrevmark, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.

**Comment by livecode-vulcan** [comment]
💙 review by @livecodefraser ok d0fcd78b95bc2a27cc7cd5f4b9de8f84855d5992

**Comment by livecode-vulcan** [comment]
😎 test success d0fcd78b95bc2a27cc7cd5f4b9de8f84855d5992
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/228)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/227)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/228)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/223)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/230)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/226)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/231)

**Comment by livecodefraser** [comment]
I'm curious as to how this managed to build when there are merge conflicts...

**Comment by runrevmark** [comment]
Hmmm - that is puzzling! Maybe it built before the is really stuff was integrated (the conflict was due to parseerrors.h)

**Comment by livecodefraser** [comment]
@livecode-vulcan review okay 622a8e4

**Comment by livecode-vulcan** [comment]
💙 review by @livecodefraser ok 622a8e45eefcc1dbfc24c3a1c334b07661bfe69d

**Comment by livecode-vulcan** [comment]
😎 test success 622a8e45eefcc1dbfc24c3a1c334b07661bfe69d
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/245)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/243)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/245)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/240)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/247)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/242)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/248)

**Pull request merged**